src/coreclr/jit/gentree.cpp

```diff
@@ -15671,14 +15671,25 @@ GenTree* Compiler::gtFoldExprConst(GenTree* tree)
                     break;
 
                 case GT_ADD:
+                {
                     itemp = i1 + i2;
                     if (tree->gtOverflow() && CheckedOps::AddOverflows(INT32(i1), INT32(i2), tree->IsUnsigned()))
                     {
                         goto INTEGRAL_OVF;
                     }
-                    i1       = itemp;
-                    fieldSeq = GetFieldSeqStore()->Append(op1->AsIntCon()->gtFieldSeq, op2->AsIntCon()->gtFieldSeq);
+                    i1 = itemp;
+
+                    FieldSeq* op1FieldSeq = op1->AsIntCon()->gtFieldSeq;
+                    FieldSeq* op2FieldSeq = op2->AsIntCon()->gtFieldSeq;
+
+                    if ((op1FieldSeq != nullptr) && (op2FieldSeq != nullptr))
+                    {
+                        // We cannot combine two field sequences
+                        return nullptr;
+                    }
+                    fieldSeq = GetFieldSeqStore()->Append(op1FieldSeq, op2FieldSeq);
```

Comment:
What are the trees and circumstances when this is hit? It is not supposed to happen.

Reply:
The assertion was in x86 windows:
```
[02:40:33] ISSUE: <ASSERT> #416319 D:\a\_work\1\s\src\coreclr\jit\gentree.cpp (19329) - Assertion failed '!"Duplicate field sequences!"' in 'add_i._add:main():int' during 'Importation' (IL size 69; hash 0x8b97301b; Tier0)
```

For what appeared to be: https://github.com/dotnet/runtime/blob/main/src/tests/JIT/IL_Conformance/Old/Conformance_Base/add_i.il

The IL does several additions of static field addresses with constants and static field addresses with other static field addresses, such as:
```
ldc.i4    0x000000FF
ldsflda    int32 add_i._add::PTR
add
```
or
```
ldsflda    int32 add_i._add::PTR
conv.i4
ldsflda    int32 add_i._add::PTR
conv.i4
add
```

It wasn't being hit for 64-bit

Reply:
Updated to change the assert instead, as we discussed through on Discord:
```diff
-    assert(!"Duplicate field sequences!");
+    // In UB-like code (such as manual IL) we can see an addition of two static field addresses.
+    // Treat that as cancelling out the sequence, since the result will point nowhere.
+    //
+    // It may be possible for the JIT to encounter other types of UB additions, such as due to
+    // complex optimizations, inlining, etc. In release we'll still do the right thing by returning
+    // nullptr here, but the more conservative assert can help avoid JIT bugs
+
+    assert(a->GetKind() == FieldKind::SimpleStaticKnownAddress);
+    assert(b->GetKind() == FieldKind::SimpleStaticKnownAddress);
```